File: apps/llm-gateway/cocola_llm_gateway/auth/jwt.py
"""Minimal, dependency-free HS256 JWT (sign + verify).

Why hand-rolled instead of PyJWT? The gateway already runs hermetic, stdlib-only
tests (network forbidden, no bound port). A signed token only needs HMAC-SHA256
over base64url(header).base64url(payload) — that is ~40 lines of stdlib and adds
no third-party surface to audit. If we later need RS256/JWKS (multi-issuer), this
is the single module to swap; nothing else imports `hmac`/`hashlib` for auth.

Tokens are compact JWS (JSON Web Signature), the standard three-segment form:
    base64url(header) "." base64url(payload) "." base64url(signature)

This module is intentionally tiny and pure: no I/O, no clock injection beyond an
optional `now` for tests. Claims we use:
    sub  -> user_id      (employee id / username)
    ten  -> tenant_id    (team/department; "" means the default tenant)
    iat  -> issued-at    (unix seconds)
    exp  -> expiry       (unix seconds; optional — omit for non-expiring tokens)
    iss  -> issuer       (defaults to "cocola")
    jti  -> token id     (per-token id; the revocation denylist key)
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from dataclasses import dataclass

_ALG = "HS256"


class JWTError(Exception):
    """Raised on any malformed/invalid/expired token. Never leaks the secret."""


def _b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def _b64url_decode(seg: str) -> bytes:
    # JWT strips '=' padding; restore it before decoding.
    pad = "=" * (-len(seg) % 4)
    try:
        return base64.urlsafe_b64decode(seg + pad)
    except Exception as e:  # noqa: BLE001 - normalize to JWTError
        raise JWTError("malformed base64url segment") from e


def _sign(signing_input: bytes, secret: str) -> str:
    sig = hmac.new(secret.encode("utf-8"), signing_input, hashlib.sha256).digest()
    return _b64url_encode(sig)


def encode(payload: dict, secret: str) -> str:
    """Sign a claims dict into a compact JWS string."""
    if not secret:
        raise JWTError("signing secret must not be empty")
    header = {"alg": _ALG, "typ": "JWT"}
    h = _b64url_encode(json.dumps(header, separators=(",", ":")).encode("utf-8"))
    p = _b64url_encode(json.dumps(payload, separators=(",", ":")).encode("utf-8"))
    signing_input = f"{h}.{p}".encode("ascii")
    s = _sign(signing_input, secret)
    return f"{h}.{p}.{s}"


def decode(token: str, secret: str, *, now: float | None = None) -> dict:
    """Verify signature + expiry and return the claims dict.

    Raises JWTError on any problem (bad shape, wrong alg, bad signature, expired).
    Signature comparison is constant-time.
    """
    if not secret:
        raise JWTError("verifying secret must not be empty")
    parts = token.split(".")
    if len(parts) != 3:
        raise JWTError("token must have three segments")
    h_seg, p_seg, sig_seg = parts

    try:
        header = json.loads(_b64url_decode(h_seg) or b"{}")
    except Exception as e:  # noqa: BLE001
        raise JWTError("header is not valid JSON") from e
    if not isinstance(header, dict) or header.get("alg") != _ALG:
        raise JWTError(f"unsupported alg {header.get('alg')!r}; expected {_ALG}")

    signing_input = f"{h_seg}.{p_seg}".encode("ascii")
    expected = _sign(signing_input, secret)
    if not hmac.compare_digest(expected, sig_seg):
        raise JWTError("signature mismatch")

    try:
        claims = json.loads(_b64url_decode(p_seg))
    except Exception as e:  # noqa: BLE001
        raise JWTError("payload is not valid JSON") from e
    if not isinstance(claims, dict):
        raise JWTError("payload must be a JSON object")

    exp = claims.get("exp")
    if exp is not None:
        t = time.time() if now is None else now
        if t >= float(exp):
            raise JWTError("token expired")
    return claims


@dataclass(frozen=True, slots=True)
class Identity:
    """Resolved caller identity extracted from a verified token.

    `user_id` is the billing + quota subject. `tenant_id` groups users for an
    optional second quota layer; "" means the default/unset tenant.
    """

    user_id: str
    tenant_id: str = ""
    issued_at: float = 0.0
    expires_at: float | None = None
    # token_id is the `jti` claim: the per-token id used to consult the
    # revocation denylist. "" when the token predates jti or auth is disabled.
    token_id: str = ""

    @property
    def is_authenticated(self) -> bool:
        return bool(self.user_id)

File: packages/go-common/token/jwt.go
// Package token mints and inspects the cocola-signed HS256 JWTs that ARE the
// Claude Agent SDK's ANTHROPIC_API_KEY. It is shared go-common code: admin-api
// mints tokens with it and the gateway BFF verifies them with the SAME codec,
// so there is exactly one HS256 implementation on the Go side (no third
// hand-rolled copy). This is the Go counterpart of the Python gateway's
// auth/jwt.py: the wire format (compact JWS, base64url without padding, header
// {"alg":"HS256","typ":"JWT"}, claims sub/ten/iat/exp/iss) is byte-compatible,
// so a token minted here verifies in the gateway and vice versa. The
// cross-language interop e2e proves it.
//
// Why hand-rolled HS256 again (instead of a JWT library)? Symmetric signing of
// compact JWS is a few lines of crypto/hmac+sha256; matching the Python side
// exactly is easier when both are deliberately tiny and we own every byte. If
// we ever move to RS256/JWKS, this is the single file to swap on the Go side.
package token

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"time"
)

const alg = "HS256"

// ErrInvalid is returned for any malformed/invalid/expired token. It never
// embeds the secret or signature so it is safe to surface to clients.
var ErrInvalid = errors.New("token: invalid")

// Claims are the fields cocola puts in the signed token. The short keys match
// the Python gateway exactly (sub/ten/iat/exp/iss).
type Claims struct {
	Subject  string `json:"sub"`
	Tenant   string `json:"ten,omitempty"`
	Email    string `json:"eml,omitempty"`
	Name     string `json:"nam,omitempty"`
	Username string `json:"usr,omitempty"`
	IssuedAt int64  `json:"iat,omitempty"`
	Expires  int64  `json:"exp,omitempty"`
	Issuer   string `json:"iss,omitempty"`
	// ID is the JWT ID (jti): a per-token opaque identifier. It is the key the
	// admin-api's revocation denylist uses and the gateway reads back from a
	// verified token to check "has this specific token been revoked?". Without
	// it, a leaked-but-unexpired token could not be killed before exp.
	ID string `json:"jti,omitempty"`
}

type header struct {
	Alg string `json:"alg"`
	Typ string `json:"typ"`
}

func b64encode(raw []byte) string { return base64.RawURLEncoding.EncodeToString(raw) }

func b64decode(seg string) ([]byte, error) {
	b, err := base64.RawURLEncoding.DecodeString(seg)
	if err != nil {
		return nil, ErrInvalid
	}
	return b, nil
}

func sign(signingInput, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(signingInput))
	return b64encode(mac.Sum(nil))
}

// Encode signs a claims set into a compact JWS string.
func Encode(c Claims, secret string) (string, error) {
	if secret == "" {
		return "", ErrInvalid
	}
	hb, err := json.Marshal(header{Alg: alg, Typ: "JWT"})
	if err != nil {
		return "", err
	}
	pb, err := json.Marshal(c)
	if err != nil {
		return "", err
	}
	signingInput := b64encode(hb) + "." + b64encode(pb)
	return signingInput + "." + sign(signingInput, secret), nil
}

// Decode verifies the signature + expiry and returns the claims. now is unix
// seconds; pass 0 to use the wall clock. Signature comparison is constant time.
func Decode(tok, secret string, now int64) (Claims, error) {
	var zero Claims
	if secret == "" {
		return zero, ErrInvalid
	}
	parts := strings.Split(tok, ".")
	if len(parts) != 3 {
		return zero, ErrInvalid
	}
	hb, err := b64decode(parts[0])
	if err != nil {
		return zero, ErrInvalid
	}
	var h header
	if err := json.Unmarshal(hb, &h); err != nil || h.Alg != alg {
		return zero, ErrInvalid
	}
	signingInput := parts[0] + "." + parts[1]
	if !hmac.Equal([]byte(sign(signingInput, secret)), []byte(parts[2])) {
		return zero, ErrInvalid
	}
	pb, err := b64decode(parts[1])
	if err != nil {
		return zero, ErrInvalid
	}
	var c Claims
	if err := json.Unmarshal(pb, &c); err != nil {
		return zero, ErrInvalid
	}
	if c.Expires != 0 {
		t := now
		if t == 0 {
			t = time.Now().Unix()
		}
		if t >= c.Expires {
			return zero, ErrInvalid
		}
	}
	return c, nil
}
